Tutorial for adiabatic sweep (#36)
qutip/qutip-julia-tutorials: QuantumToolbox.jl/time_evolution/adiabatic.qmd

---
title: "Adiabatic sweep (with `QuantumObjectEvolution`)"
author: Li-Xun Cai
date: 2025-04-09  # last update (keep this comment as a reminder)

engine: julia
---

Inspirations taken from [the QuTiP tutorial](https://nbviewer.org/urls/qutip.org/qutip-tutorials/tutorials-v5/lectures/Lecture-8-Adiabatic-quantum-computing.ipynb) by J. R. Johansson.

This tutorial mainly demonstrates the use of [`QuantumObjectEvolution`](https://qutip.org/QuantumToolbox.jl/stable/resources/api#QuantumToolbox.QuantumObjectEvolution). 

## Introduction

> The quantum adiabatic theorem is a fundamental principle in quantum mechanics that describes how quantum systems evolve when subjected to time-dependent conditions. This theorem states that if a quantum system starts in an eigenstate (typically the ground state) of its initial Hamiltonian, and if the Hamiltonian changes sufficiently slowly, the system will remain in the corresponding eigenstate of the evolving Hamiltonian throughout the process. 

Reviewer: ```suggestion
The quantum adiabatic theorem is a fundamental principle in quantum mechanics that describes how quantum systems evolve when subjected to time-dependent conditions. This theorem states that if a quantum system is initialized in an eigenstate (typically the ground state) of its initial Hamiltonian, and if the Hamiltonian changes sufficiently slow, the system will remain in the corresponding eigenstate of the evolving Hamiltonian throughout the process. 
```

> For this theorem to hold, certain conditions must be satisfied: the system must evolve slowly enough compared to the energy gap between the relevant eigenstate and other energy levels, and this gap must remain non-zero throughout the evolution.

Reviewer: ```suggestion
For this theorem to hold, certain conditions must be satisfied: the system must evolve slow enough compared to the energy gap between the relevant eigenstate and other energy levels, and this gap must remain non-zero throughout the evolution.
```

> Here, we will demonstrate the well-known application of the adiabatic theorem in quantum computing called \emph{adiabatic sweep}. This is a method for preparing some desired quantum state by slowly evolving a quantum system with a time-dependent Hamiltonian. Essentially, the adiabatic theorem allows us to prepare the ground state of the final Hamiltonian $H_1$ from the ground state of the initial Hamiltonian $H_0$.

Reviewer: ```suggestion
Here, we will demonstrate the well-known application of the adiabatic theorem in quantum computing called $\emph{adiabatic sweep}$. This is a method for preparing some desired quantum state by slowly evolving a quantum system with a time-dependent Hamiltonian. Essentially, the adiabatic theorem allows us to prepare the ground state of the final Hamiltonian $H_1$ from the ground state of the initial Hamiltonian $H_0$.
```

## Model

> We consider a chain of $N$ identical spin-half particles to study their spin dynamics and set our interest to finding the ground state of the condition that the spin chain has some random gap, leading to the random magnitude $g_i$ of $\hat{\sigma}^i_x \hat{\sigma}^{i+1}_x$ interaction with the neighboring spin.

Reviewer: ```suggestion
We consider a chain of $N$-identical spin-$1/2$ particles to study their spin dynamics and set our interest to find the ground state of the condition that the spin chain has some random gap, leading to the random magnitude $g_i$ of $\hat{\sigma}^i_x \hat{\sigma}^{i+1}_x$ interaction with the neighboring spin.
```

Initially, we prepare the system such that the spins are free from interaction with each other and are all in the ground state, i.e.,

$$H_0 = \sum_i^N \frac{\varepsilon_0}{2} \hat{\sigma}^i_z,$$

> $$|\psi(0)\rangle = \otimes_{i=1}^N |g\rangle$$

Reviewer: ```suggestion
$$|\psi(0)\rangle = \bigotimes_{i=1}^N |g\rangle$$
```

Then, gradually, the Hamiltonian evolves to:

$$H_1 = \sum_{i=1}^N \frac{\varepsilon_0}{2} \hat{\sigma}^i_z + \sum_{i=1}^{N-1} g_i \hat{\sigma}^i_x \hat{\sigma}^{i+1}_x,$$

whose ground state are desired. By gradual change, we are subject to the simplest form of adiabatic sweep, i.e.,

> $$H(t,T) = H_0 * (1-t/T) + H_1 * t/T,$$
> 
> where the parameter $T$, the time for full evolution, determines how slow the Hamiltonian changes.

Reviewer: ```suggestion
$$H(t,T) = \left(1-\frac{t}{T}\right) H_0 + \frac{t}{T} H_1,$$

where the parameter $T$ determines how slow the Hamiltonian changes in time.
```

## Code demonstration

```{julia}
using QuantumToolbox
using CairoMakie
```

```{julia}
N = 8 # number of spins
ε0 = 1 # energy gap of the spins

gs = rand(N-1) # use rand function for the random interaction strengths
```

```{julia}

> Is = Vector{QuantumObject}(fill(qeye(2), N));

Reviewer 2: Why do we need to specify this is a vector? Why don't we just do

```suggestion
Is = fill(qeye(2), N);
```

Author: Because `qeye(2).data` is a diagonal matrix and `fill(qeye(2),N)` is then a Vector of diagonal, which cannot take `sigmax()` as its element.

`multisite_operator` is better anyway so I will adapt to that later.

H0 = QuantumObject(
    zeros((2^N, 2^N)),
    Operator,
    Tuple(fill(2, N))
)

for idx in 1:N
    is = copy(Is)
    is[idx] = ε0/2 * sigmaz()
    H0 += kron(is...)
end

ψ0 = kron(fill(basis(2,1), N)...)

> print(H0)

Reviewer 2: The matrix is dense and not sparse. This is not optimal. I think this can be fixed by defining `H0` with `spzeros` instead of `zeros`.

I would rewrite everything using the [`multisite_operator`](https://qutip.org/QuantumToolbox.jl/stable/resources/api#QuantumToolbox.multisite_operator) function. Something like

```julia
H0 = sum(i -> ε0/2 * multisite_operator(Val(N), i=>sigmaz()), 1:N)
```

```

```{julia}

> H1 = QuantumObject(
>     zeros((2^N, 2^N)),
>     Operator,
>     Tuple(fill(2, N))
> )
> 
> for idx in 1:N
>     is = copy(Is)
>     is[idx] = ε0/2 * sigmaz()
>     H1 += kron(is...)
> end
> 
> for idx in 1:N-1
>     is = copy(Is)
>     is[idx] = gs[idx] * sigmax()
>     is[idx+1] = sigmax()
> 
>     h_int = kron(is...)
>         
>     H1 += h_int
> end
> 
> print(H1)

Reviewer 2: Same think for H1 here. I think we can use

```suggestion
H1 = sum(i -> ε0/2 * multisite_operator(Val(N), i=>sigmaz()), 1:N) + sum(i -> gs[i] * multisite_operator(Val(N), i=>sigmax(), i+1=>sigmax()), 1:N-1)
```

```

Here, we define the function `H(T)` where `T` is the total evolution time.
```{julia}

> H(T) = QuantumObjectEvolution((
>     (H0, (p,t) -> 1 - t/T),  # the dummy parameter `p` is not used but still required by the ODE solver
>     (H1, (p,t) -> t/T),
> ))

Reviewer 2: Instead of put it as a function, you can define it using the parameters

```suggestion
H = QuantumObjectEvolution((
     (H0, (p,t) -> 1 - t/p.T), 
     (H1, (p,t) -> t/p.T),
 ))
```

So that we can show the potentiality of `QuantumObjectEvolution`

```

```{julia}
function ψg(H)
    _, vecs = eigenstates(H)
    return vecs[1]
end

ψf_truth = ψg(H1) |> to_sparse
print(ψf_truth)
```

We can see that the truthful ground state we are preparing is indeed very complex. 

For the adiabatic theorem to apply, we have to check for the gap between the ground state and first excited state remaining non-zero throughout the evolution.
```{julia}
T = 10
tlist = 0:0.1:T

> eigs = Array{Real}(undef, length(tlist), 2^N)

Reviewer 2: `Real` is an abstract type. Better to specify it

```suggestion
eigs = Array{Float64}(undef, length(tlist), 2^N)
```

> for (idx, t) in enumerate(tlist)
>     vals, _ = eigenstates(H(T)(t))
>     eigs[idx,:] = vals
> end

Reviewer 2: And here it becomes

```suggestion
params = (T = T, )
for (idx, t) in enumerate(tlist)
     vals, _ = eigenstates(H(params, t))
     eigs[idx,:] = vals
 end
```

```

```{julia}

> fig = Figure()

Reviewer 2: Maybe we can define the size of the figure

```suggestion
fig = Figure(size=(800, 400))
```

ax = Axis(fig[1,1], xticks = (0:0.25:1, ["$(t)T" for t in 0:0.25:1]))

for idx in 1:20 # only check for the lowest 20 eigenvalues
    color = (idx == 1) ? :magenta : (:gray,0.5)
    lines!(ax, range(0,1,length(tlist)), eigs[:,idx], label = string(idx), color = color)
end

display(fig)
```

The plot shows that the gap is nonvanishing and thus validates the evolution.
```{julia}
Tlist = 10 .^ (0:0.25:1.25)
results = map(Tlist) do T 
    tlist = range(0,T, 101)
    sesolve(H(T), ψ0, tlist, e_ops = [H1, ket2dm(ψf_truth)]) 
    # check the expectation value with respect to the final Hamiltonian and
    # the fidelity to the truthful ground state throughout the evolution

> end;

Reviewer 2: And here

```julia
results = map(Tlist) do T 
     tlist = range(0,T, 101)
     params = (T = T,)
     sesolve(H, ψ0, tlist, e_ops = [H1, ket2dm(ψf_truth)], params=params) 
     # check the expectation value with respect to the final Hamiltonian and
     # the fidelity to the truthful ground state throughout the evolution
 end;
```
